form/src/main/java/org/dhis2/form/ui/provider/FieldProvider.kt

```diff
@@ -0,0 +1,162 @@
+package org.dhis2.form.ui.provider
+
+import android.content.Context
+import android.text.TextWatcher
+import android.view.ContextThemeWrapper
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.viewinterop.AndroidViewBinding
+import androidx.databinding.DataBindingUtil
+import androidx.databinding.ViewDataBinding
+import org.dhis2.form.BR
+import org.dhis2.form.R
+import org.dhis2.form.model.FieldUiModel
+import org.dhis2.form.ui.LatitudeLongitudeTextWatcher
+import org.dhis2.form.ui.event.RecyclerViewUiEvents
+import org.dhis2.form.ui.intent.FormIntent
+import org.hisp.dhis.android.core.common.ValueType
+import org.hisp.dhis.mobile.ui.designsystem.component.InputShellState
+import org.hisp.dhis.mobile.ui.designsystem.component.InputText
+import org.hisp.dhis.mobile.ui.designsystem.component.LegendData
+import org.hisp.dhis.mobile.ui.designsystem.component.SupportingTextData
+import org.hisp.dhis.mobile.ui.designsystem.component.SupportingTextState
+
+@Composable
+internal fun FieldProvider(
+    modifier: Modifier,
+    context: Context,
+    fieldUiModel: FieldUiModel,
+    needToForceUpdate: Boolean,
+    textWatcher: TextWatcher,
+    coordinateTextWatcher: LatitudeLongitudeTextWatcher,
+    uiEventHandler: (RecyclerViewUiEvents) -> Unit,
+    intentHandler: (FormIntent) -> Unit,
+) {
+    when {
+        fieldUiModel.optionSet == null && fieldUiModel.valueType == ValueType.TEXT -> {
```

Comment:
maybe it's a good idea to extract the fieldUiModel.optionSet null check so that we don't have to check every time we add a value type?

Reply:
optionSet will work for InputRadioButton, InputCheckbox, InputMatrix and InputSequential. Any way, I agree this when should be rethinked